File: Tic_Tac_Toe_AI/python code/ai.py
from game import Game
import random

def f(board):
    l = []
    for i in board:
        l.append(tuple(i))
    return tuple(l)

class Player:
    def __init__(self, n):
        self.name = n
        self.q_table = {}

    def rand_play(self, game):
        v = []
        for i in range(3):
            for j in range(3):
                if game.board[i][j] == '#':
                    v.append((i,j))

        if len(v) != 0:
            index = random.randint(0, len(v)-1)
            row, column = v[index]
            game.play(row, column, self.name)

    def next_moves(self, game):
        board = game.board.copy()
        v = []
        for i in range(3):
            for j in range(3):
                if game.board[i][j] == '#':
                    board[i][j] = self.name
                    v.append(board)
                    board = game.board.copy()
        return v

    def other_player(self):
        if self.name == 'x':
            return 'o'
        else:
            return 'x'

    def train_play(self, game):
        moves = self.next_moves(game)

        for b in moves:
            game2 = Game()
            game2.board = b
            if game2.winner() == self.name:
                self.q_table[f(b)] = 1
                self.q_table[f(game.board)] = 1
            elif game2.winner() == self.other_player():
                self.q_table[f(b)] = -1
            elif game2.isgameover() and game2.winner() == '#':
                self.q_table[f(b)] = -1
            else:
                val = self.q_table.get(f(b))
                self.q_table[f(b)] = 0 if val == None else val

        max = -1
        for b in moves:
            a = self.q_table.get(f(b))
            a = 0 if a == None else a
            if a > max:
                max = a

        self.q_table[f(game.board)] = max

        game.board = moves[random.randint(0, len(moves)-1)]

    def play(self, game):
        moves = self.next_moves(game)
        if len(moves) == 1:
            game.board = moves[0]
        else:
            board = moves[0]
            val = self.q_table.get(f(board))
            val = 0 if val == None else val
            for i in range(1, len(moves) - 1):
                temp_board = moves[i]
                temp_val = self.q_table.get(f(temp_board))
                temp_val = 0 if temp_val == None else temp_val
                if temp_val > val:
                    val = temp_val
                    board = temp_board
                
            game.board = board

File: Tic_Tac_Toe_AI/src/ai.rs
use crate::Game;
use rand::Rng;
use std::collections::HashMap;

pub struct Player {
    name:char,
    #[allow(unused)]
    pub q_table:HashMap<[[char;3];3], i32>
}

impl Player {
    pub fn new(n:char) -> Self {
        Self {name:n, q_table:HashMap::new()}
    }

    #[allow(unused)]
    pub fn rand_play(&self, game:&mut Game) {
        let mut r = rand::thread_rng();
        let mut v = Vec::new();
        for i in 0..3 {
            for j in 0..3 {
                if game.board[i][j] == '#' {
                    v.push((i,j));
                }
            }
        }
        
        if v.len() != 0 {
            let index = r.gen_range(0..v.len());
            let (row, column) = v[index];
            game.play(row, column, self.name);
        }
    }

    #[allow(unused)]
    fn next_moves(&self, game:&Game) -> Vec<[[char; 3]; 3]> {
        let mut board = game.board.clone();
        let mut v = Vec::new();
        for i in 0..3 {
            for j in 0..3 {
                if game.board[i][j] == '#' {
                    board[i][j] = self.name;
                    v.push(board);
                    board = game.board.clone();
                }
            }
        }
        v
    }

    #[allow(unused)]
    fn other_player(&self) -> char {
        if self.name == 'x' {
            'o'
        } else {
            'x'
        }
    }

    #[allow(unused)]
    pub fn train_play(&mut self, game:&mut Game) {
        let moves = self.next_moves(game);

        for b in &moves {
            let game2 = Game{board:*b};
            if game2.winner() == self.name {
                self.q_table.insert(*b, 1);
                self.q_table.insert(game.board, 1);
            } else if game2.winner() == self.other_player() {
                self.q_table.insert(*b, -1);
            } else if game2.isgameover() && game2.winner() == '#' {
                self.q_table.insert(*b, -1);
            } else {
                let val = match self.q_table.get(b) {
                    Some(a) => a,
                    None => &0
                };
                self.q_table.insert(*b, *val);
            }
        }

        let mut max = -1;
        for b in &moves {
            let a = self.q_table.get(b).unwrap_or(&0);
            if *a > max {
                max = *a
            }
        }

        self.q_table.insert(game.board, max);

        let mut r = rand::thread_rng();
        game.board = moves[r.gen_range(0..moves.len())];

    }

    #[allow(unused)]
    pub fn play(&self, game:&mut Game) {
        let moves = self.next_moves(game);
        if moves.len() == 1 {
            game.board = moves[0];
        } else {
            let mut board = moves[0];
            let mut val = self.q_table.get(&board).unwrap_or(&0);
            for i in 1..moves.len() {
                let temp_board = moves[i];
                let temp_val = self.q_table.get(&temp_board).unwrap_or(&0);
                if temp_val > val {
                    val = temp_val;
                    board = temp_board;
                }
            }
            game.board = board;
        }
    }
}